test_runner, cli: add --test-timeout flag

shubham9411 opened the pull request:
This commit introduces a new command-line argument, `--test-timeout`, which allows users to specify a timeout for the test runner CLI.
Fixes: https://github.com/nodejs/node/issues/50431 
<!--
Before submitting a pull request, please read:

- the CONTRIBUTING guide at https://github.com/nodejs/node/blob/HEAD/CONTRIBUTING.md
- the commit message formatting guidelines at
  https://github.com/nodejs/node/blob/HEAD/doc/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

If you believe this PR should be highlighted in the Node.js CHANGELOG
please add the `notable-change` label.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

MoLow commented on test/fixtures/test-runner/timeout.js:
```diff
@@ -0,0 +1,15 @@
+  stream.on('test:pass', common.mustNotCall());
+  // eslint-disable-next-line no-unused-vars
+  for await (const _ of stream);
+});
```
```suggestion
});

```

shubham9411 commented:
Can anyone help me with the failed tests? I'm not sure if they're flaky or if I need to fix something.

MoLow commented:
CI seems to fail due to a zombie process:
```
12:06:01 ps awwx | grep Release/node | grep -v grep | cat
12:06:01 69416   ??  R      8:24.87 /Users/iojs/build/workspace/node-test-commit-osx-arm/nodes/osx11/out/Release/node --test-timeout 50 /Users/iojs/build/workspace/node-test-commit-osx-arm/nodes/osx11/test/fixtures/test-runner/never_ending_sync.js
```

MoLow commented on test/fixtures/test-runner/timeout.js:
```diff
@@ -0,0 +1,15 @@
+const common = require('../../common');
+const fixtures = require('../../common/fixtures');
+
+test('should support timeout', async () => {
+  const stream = test.run({files: [
```
```suggestion
test('should support timeout', async ({ signal }) => {
  const stream = test.run({
  signal,
  files: [
```

this will kill the zombie processes

shubham9411 replied:
Updated thanks.

MoLow commented:
@shubham9411 this is still happening:
https://ci.nodejs.org/job/node-test-commit-osx-arm/14241/nodes=osx11/console
```
08:35:34 All tests passed.
08:35:34 ./node ./test/embedding/test-embedding.js
08:35:35 ps awwx | grep Release/node | grep -v grep | cat
08:35:35 39342   ??  R      8:20.17 /Users/iojs/build/workspace/node-test-commit-osx-arm/nodes/osx11/out/Release/node --test-timeout 50 /Users/iojs/build/workspace/node-test-commit-osx-arm/nodes/osx11/test/fixtures/test-runner/never_ending_sync.js
```

shubham9411 commented:
@MoLow this looks like `never_ending_sync` is being called without `--test` flag and so test-timeout is not cancelling it. I am not sure why it is behaving like this, I have used `spawnSync` with both `--test` and `--test-timeout` args. I'll check more why it is happening.

shubham9411 commented:
Hi @MoLow, I have updated the test, I have added a timeout to `spawnSync` as well so the process will be killed if called without `--test`. Still not sure why it is being called without `--test` flag (not reproducing in my system).

shubham9411 commented:
Hi @MoLow, Can you please review the PR and let me know if anything else is needed. thanks

MoLow commented:
please remove the merge commits by squashing or rebasing

shubham9411 commented:
updated

MoLow commented:
@shubham9411 there are relevant failures:
https://ci.nodejs.org/job/node-test-commit-linux-containered/40226/nodes=ubuntu1804_sharedlibs_shared_x64/console
https://ci.nodejs.org/job/node-test-commit-linux/54697/nodes=alpine-last-latest-x64/console

```
11:37:47 not ok 2461 parallel/test-runner-cli-timeout
11:37:47   ---
11:37:47   duration_ms: 1311.27600
11:37:47   severity: fail
11:37:47   exitcode: 1
11:37:47   stack: |-
11:37:47     TAP version 13
11:37:47     # Subtest: test-timeout flag
11:37:47     not ok 1 - test-timeout flag
11:37:47       ---
11:37:47       duration_ms: 522.560499
11:37:47       location: '/home/iojs/build/workspace/node-test-commit-linux-containered/test/parallel/test-runner-cli-timeout.js:7:1'
11:37:47       failureType: 'testCodeFailure'
11:37:47       error: |-
11:37:47         Expected values to be strictly equal:
11:37:47         
11:37:47         null !== 1
11:37:47         
11:37:47       code: 'ERR_ASSERTION'
11:37:47       name: 'AssertionError'
11:37:47       expected: 1
11:37:47       operator: 'strictEqual'
11:37:47       stack: |-
11:37:47         TestContext.<anonymous> (/home/iojs/build/workspace/node-test-commit-linux-containered/test/parallel/test-runner-cli-timeout.js:11:10)
11:37:47         Test.runInAsyncScope (node:async_hooks:206:9)
11:37:47         Test.run (node:internal/test_runner/test:631:25)
11:37:47         Test.start (node:internal/test_runner/test:542:17)
11:37:47         startSubtest (node:internal/test_runner/harness:216:17)
11:37:47       ...
11:37:47     1..1
11:37:47     # tests 1
11:37:47     # suites 0
11:37:47     # pass 0
11:37:47     # fail 1
11:37:47     # cancelled 0
11:37:47     # skipped 0
11:37:47     # todo 0
11:37:47     # duration_ms 583.480686
```

shubham9411 commented:
@MoLow, Can you please request ci for this pr? thanks

MoLow commented:
Congrats @shubham9411 🎉

shubham9411 commented:
Thank you @MoLow 😄
